src/gasEstimator/utils.ts

```diff
@@ -0,0 +1,122 @@
+import { BigNumberish, BigNumber, getDefaultProvider } from 'ethers';
+import { BigNumber as BigNumberJs } from 'bignumber.js';
+import { RelayHub__factory } from '@rsksmart/rif-relay-contracts';
+import type { EnvelopingRequest } from '../common/relayRequest.types';
+import { isDeployRequest } from '../common/relayRequest.utils';
+import type { EnvelopingTxRequest } from '../common/relayTransaction.types';
+import RelayClient from '../RelayClient';
+import {
+  applyGasCorrectionFactor,
+  ESTIMATED_GAS_CORRECTION_FACTOR,
+} from '../utils';
+
+const standardMaxPossibleGasEstimation = async (
+  { relayRequest, metadata: { signature } }: EnvelopingTxRequest,
+  relayWorkerAddress: string,
+  tokenEstimation: BigNumber,
+  estimatedGasCorrectionFactor?: BigNumberish
+): Promise<BigNumber> => {
+  const {
+    request,
+    relayData: { gasPrice },
+  } = relayRequest;
+
+  const provider = getDefaultProvider();
+
+  const relayHub = RelayHub__factory.connect(
+    request.relayHub.toString(),
+    provider
+  );
+
+  const methodToEstimate = isDeployRequest(relayRequest)
+    ? await relayHub.populateTransaction.deployCall(relayRequest, signature)
+    : await relayHub.populateTransaction.relayCall(relayRequest, signature);
+
+  const relayEstimation = await provider.estimateGas({
+    ...methodToEstimate,
+    gasPrice,
+    from: relayWorkerAddress,
+  });
+
+  const correctedEstimation = applyGasCorrectionFactor(
+    relayEstimation,
+    estimatedGasCorrectionFactor
+  );
+
+  return tokenEstimation.add(correctedEstimation);
+};
+
+const linearFitMaxPossibleGasEstimation = async (
+  envelopingRequest: EnvelopingRequest,
+  tokenEstimation: BigNumber,
+  internalEstimationCorrection?: BigNumberish,
+  estimatedGasCorrectionFactor?: BigNumberish
+): Promise<BigNumber> => {
+  if (isDeployRequest(envelopingRequest)) {
+    throw Error('LinearFit estimation not implemented for deployments');
+  }
+
+  const {
+    request: { data, from, to },
+    relayData: { gasPrice },
+  } = envelopingRequest;
+
+  const relayClient = new RelayClient();
```

Comment:
By doing this here I think that we are creating a dependency on the relayClient and the constructor change, we will need to include those changes here or even pass the client as parameter.

Reply:
does this mean we're going to introduce a circular dependency?

Reply:
not circular but we are coupling this method with the relayClient.

Reply:
I think we can simplify the interaction later on. I'd leave it as it is for now